### ISSUE TITLE
[Feature request] Customizable control over when a selector triggers re-render

### ISSUE THREAD

**Issue by BenjaBobs** [facebookexperimental/Recoil#1416]
# Proposal

I would like to be able to specify a custom function for cache comparison on an `atom`/`selector` so I can tweak the behaviour.
Let's say we have a configurable algorithm with multiple steps, and some options that pertain to a step each.
```ts
const algorithmOptions = atom({
	key: 'algorithmOptions',
	default: {
		featureA: true,
		featureB: true,
		featureC: false,
	},
});
```
Now, only step 2 options are relevant for the step 2 of the algorithm, so we make a selector to fetch only that
```ts
const step2Options = selector({
	key: 'step2Options',
	get: ({get}) => {
		const allOptions = get(algorithmOptions);

		return {
			allOptions.featureB,
			allOptions.featureC,
		};
	},
	// overwrite cache comparison, so we don't trigger unnecessary rerenders by modifying algorithmOptions.featureA
	cacheCompare: (old, new) => old.featureB=== new.featureB && old.featureC === new.featureC,
})
```
That way, we don't trigger an unnecessary rerender when doing `setOptions({ ...options, featureA: !options.featureA })`.

# Motivation

This can currently be done using an intermediate `selector`, but it's a lot of boiler plate and makes the code look clunky.
Here's a code sandbox showing the intermediate `selector` pattern: https://codesandbox.io/s/busy-cherry-u4rx6?file=/src/App.tsx

The real life use cases for this are when you get a state object from a backend, which you want to split out into multiple parts, or if you have a form for configuration of something, and you want to save the state of the form as is.

**Comment by pc-erin** [comment]
This is an urgently needed feature.
Without it any compound atom triggers a rerender even when no constituent value has changed.
If that atom is something that changes frequently, like mouse position, it has a big impact on application performance.

**Comment by mondaychen** [comment]
In the example, any reason why you can't use three separate `atom`s, or an `atomFamily` as feature flags?

**Comment by pc-erin** [comment]
Not who you were asking, but I have an example that involves manipulating Points (`{ x: number, y: number }`), and it wouldn't make much sense to store the x and y values in separate atoms since they will only ever be accessed together.

It also wouldn't help with the issue where selectors trigger rerenders even when there are no updates because the derived state is always a new object.

```typescript
export const viewportMousePositionState = atom<Point>({
	key: 'viewportMousePosition',
	default: ZERO_POINT
})

export const gridMousePositionState = selector<Point>({
	key: 'gridMousePosition',
	get: ({ get }) => {
		const gridViewportPosition = get(gridViewportPositionState)
		const viewportMousePosition = get(viewportMousePositionState)
		const gridOffset = get(gridOffsetState)
		const zoomLevel = get(zoomLevelState)

		return {
			x: ((viewportMousePosition.x - gridViewportPosition.x) / zoomLevel) - gridOffset.x,
			y: ((viewportMousePosition.y - gridViewportPosition.y) / zoomLevel) - gridOffset.y
		}
	},
	cachePolicy_UNSTABLE: {
		eviction: 'most-recent'
	}
})

export const gridSnappedMousePositionState = selector<Point>({
	key: 'gridSnappedMousePosition',
	get: ({ get }) => {
		const mousePosition = get(gridMousePositionState)

		return {
			x: Math.round(mousePosition.x),
			y: Math.round(mousePosition.y)
		}
	},
	cachePolicy_UNSTABLE: {
		eviction: 'most-recent'
	}
})
```

In this example any component that consumes `gridMousePositionState` or `gridSnappedMousePositionState` rerenders on every "tick" because the get method outputs are considered to be a new object on every run and there's no way to tell the memoizer how to know if two Point objects are equal.

**Comment by methodman7556** [comment]
I think this has been a long-standing issue with Recoil https://github.com/facebookexperimental/Recoil/issues/314 and solving it will make this the go-to state management system for React.

**Comment by drarmstr** [comment]
As was mentioned, this can currently be addressed via intermediate selectors from #314.  Though, it makes sense to improve the API for this without intermediate selectors.

API options to consider:
* custom comparator for a selector to compare it's output value to determine if it should propagate updates downstream.
* custom comparator for a selector to compare dependencies to determine if it should re-evaluate
* custom comparator provided at individual callsites, such as from hooks or selector `get()` to determine if it should re-render or re-evaluate.

**Comment by pc-erin** [comment]
@drarmstr If you're taking votes I'd go with option 1, but applied to both atoms and selectors.

Seems like it would scale the best. Any atom/selector that has lots of dependents or changes frequently can just have an `isEqual` option added to it and all dependent selectors and components automatically get the benefit of it. It also makes sense with the semantics of atoms as self-contained units of state. Who knows the significance of an atom's value better than the atom itself?

Ex:

```typescript
const viewportMousePositionState = atom<Point>({
  key: 'viewportMousePosition',
  default: ZERO_POINT,
  isEqual: arePointsEqual,
})

const gridMousePositionState = selector<Point>({
  key: 'gridMousePosition',
  get: ({ get }) => { ... },
  isEqual: arePointsEqual,
})

function arePointsEqual(a: Point, b: Point): boolean {
  return a.x === b.x && a.y === b.y
}
```

I'm not sure how you would implement solution 2 without complicating the selector options. You'd need to pass a deps array like for hooks that specifies the depended on atom/selector, along with the comparator function every time you want to use it.

Ex:
```typescript
const gridMousePositionState = selector<Point>({
  key: 'gridMousePosition',
  get: ({ get }) => {
    const gridViewportPosition = get(gridViewportPositionState)
    const viewportMousePosition = get(viewportMousePositionState)
    const gridOffset = get(gridOffsetState)
    const zoomLevel = get(zoomLevelState)
    ...
  },
  deps: [
    { state: viewportMousePositionState, isEqual: arePointsEqual },
    { state: gridViewportPositionState, isEqual: arePointsEqual },
    // etc
    // Seems like a lot more boilerplate, and I'm not sure what benefit it offers.
  ]
})
```

Similar thoughts for option 3: more boilerplate, not clear why you would need different comparators at different call sites.

Ex:
```typescript
const gridMousePositionState = selector<Point>({
  key: 'gridMousePosition',
  get: ({ get }) => {
    const gridViewportPosition = get(gridViewportPositionState, arePointsEqual)
    const viewportMousePosition = get(viewportMousePositionState, arePointsEqual)
    const gridOffset = get(gridOffsetState, arePointsEqual)
    const zoomLevel = get(zoomLevelState)
    ...
  }
})

function ExampleComponent() {
  const gridMousePosition = useRecoilValue(gridMousePositionState, arePointsEqual)
  ...
}
```

**Comment by BenjaBobs** [comment]
If it were me, I'd go for a mix of option 1 and 3.
Custom comparator at the selector/atom level, since I think this is going to be the most common use case.
And then I'd provide custom comparator as an optional parameter in `GetRecoilValue`, to override the native custom comparator of the recoil value.

**Comment by FokkeZB** [comment]
@drarmstr I *think* this aligns with a need we have, so posting here but LMK if this should be a separate request.

We (Zapier) have a need to be able to use some information from other selectors/atoms in a selector's `get`, without trigger the selector to re-evaluate when their values change.

A concrete example is that while for an API request made in a selector we need set X, we only want to re-evaluate the selector and as part of that redo the API request when Y⊂X (a subset of X) changes, or when we manually tell it to (using `useRecoilRefresher`).

I'd personally lean towards [your option](https://github.com/facebookexperimental/Recoil/issues/1416#issuecomment-984167629) three, as in:

```ts
const oneState = selector({
  key: 'one',
  get: async ({ get }) => {
    // This we need and want to "subscribe" to (register as dependency), so that we re-evaluate when its value changes
    const two = get(twoState);

    // This we need, but don't want to re-evaluate for
    const three = get(threeState, { subscribe: false });   

    return fetch(url, { data: { two, three } });
  }
});
```

Why I'd prefer this is that to me any call to `get()` represents creating a dependency on another value, so it makes sense to right there inform Recoil on whether it should indeed be considered a "hard" dependency that we subscribe to, or a "soft" one that we just need the information from.

I realize that for the original request, this still requires an intermediate selector to select either value from `algorithmOptions`, but I don't think that's a downside as I think it's more clear than an often hard to read compare function that has to deal with the values of all `get()`'s, of which some might be called conditional - making it even more complex.

On the implementation side I wonder if this is as easy as skipping the following line when `subscribe === false`:

https://github.com/facebookexperimental/Recoil/blob/4b985a32e0c56331b3e77d70e43dc77757b70924/packages/recoil/recoil_values/Recoil_selector.js#L703

**Comment by drarmstr** [comment]
@FokkeZB If you just want to read a value of an atom/selector without subscribing to it then you can use `useRecoilCallback()`.

**Comment by FokkeZB** [comment]
@drarmstr I'm aware of that, but notice how in my use case I need to read values inside of a selector's `get`, which means I cannot use hooks.

Is your suggestion to do something like this perhaps?

```ts
const oneState = selector({
  key: 'one',
  get: async ({ get, getCallback }) => {
    // This we need and want to "subscribe" to (register as dependency), so that we re-evaluate when its value changes
    const two = get(twoState);

    // This we need, but don't want to re-evaluate for
    const threeGetter = getCallback(({ snapshot }) => async () => snapshot.getPromise(threeState));
    const three = await threeGetter();

    return fetch(url, { data: { two, three } });
  }
});
```

I'll give that a try, but it feels hacky to me?

**Comment by BenjaBobs** [comment]
As I mentioned, the mix of option 1 and 3 would grant enough flexibility for everyone to be able to solve for their use cases, as well as being being a simple solution that doesn't get in your way, what with it being optional. I haven't read the code for triggering the subscribtion notifications, so I don't know if it's particularly easy to implement though, but I think it's the best solution.

**Comment by FokkeZB** [comment]
What I don't like about option 1 is that the puts the comparator at the producer, not the consumers (!) of values.

I might have multiple consumers of the same value, and some of them might want to re-evaluate when the consumed value changes, while others might not.

**Comment by BenjaBobs** [comment]
Yes, but by putting both options in the hands of the developer, with option 3 taking precendence over option 1, you can save a lot of code for the cases where you want the condition to be the same everywhere. You could absolutely achieve the same results of option 1 by using option 3 in all consumers, option 1 just saves you some work, and the worey that if a new dev is attached to the project, they'll forget to set the cache comparing function in the new place where they need to consume the state.

I have no evidence for this, but my gut feeling is that the most common use case is the one that option 1 solves, and the one that option 3 solves is more rare.

**Comment by FokkeZB** [comment]
> the worey that if a new dev is attached to the project, they'll forget to set the cache comparing function in the new place where they need to consume the state.

In contrast, I worry that when consuming a value I might be puzzled when I notice that my selector doesn't always re-evaluate when I *know* the consumed value has changed. Then only by looking at the implementation of the consumed value will I find out why.

Also technically I think option 3 is much easier to implement, and I don't think there's any use case that option 1 can solve, but option 3 can't. And since AFAIK the goal is to keep the API to a minimum, I think option 3 makes more sense.

**Comment by FokkeZB** [comment]
To make sure I understand your original question correctly; what does `old` and `new` refer to in your `cacheCompare`?

As I'm reading it this is @drarmstr's option 2, where `cacheCompare` is called every time any of `step2Options`'s dependencies their values change, with `old` holding their old values and `new` their new values. But what if `step2Options` does more `get()` calls, some of which perhaps conditionally? I don't see how we could keep `cacheCompare` readable.

I already explained why option 1 doesn't appeal to me either because in this case `algorithmOptions` would have a function to determine that a change of `featureA` would not cause **any** dependees to re-evaluate. What if I add a new selector that needs `algorithmOptions` and **does** want to be re-evaluate when `featureA` changes?

IMO option 3 is the only reasonable option, because here we determine whether a change of value from a dependency should trigger us to re-evaluate right where we create the dependency by calling `get()`.

I also think this should take the form of a complex comparator function, but simply a boolean. If only part of the value we `get()` should trigger us to re-evaluate, we can use intermediate selectors to get us both parts and set the flag on just the one that we **don't** want to trigger a re-evaluation.

Yes, this might feel boilerplaty, but IMO it keeps us closer to the data-flow graph idea. Think of option 3's boolean as controlling whether a line in the graph should be solid (triggering re-evaluation) or dashed (not triggering re-evaluation).

**Comment by BenjaBobs** [comment]
In my original post, my example was basically @drarmstr's option 1, where `old` was the previously calculated return value, and `new` is a newly calculated one.

I agree with your critique of option 2, I think it'll be too complex and difficult to control.

> I already explained why option 1 doesn't appeal to me either because in this case `algorithmOptions` would have a function to determine that a change of `featureA` would not cause **any** dependees to re-evaluate. What if I add a new selector that needs `algorithmOptions` and **does** want to be re-evaluate when `featureA` changes?

Then I think you should have your selector/component depend on `algorithmOptions` instead of `step2Options`.
I firmly believe it is `step2Options`'s responsibility to declare when it's result is a new value, or not a new value.
Option 1, as I see it, best encapsulates this responsibility.
The reason I would like like option 3, and for that to be able to override option 1, is that I think it's always good to provide as much control as possible, but to have some simple defaults.

**Comment by FokkeZB** [comment]
Ah.... I thought #314 had (also) address this example use case, but I gave it a try at https://codesandbox.io/s/recoil-playground-jkvq7?file=/pages/invalidation-option1.tsx:988-1059 and you're right that it does not. Although the [hidden](https://github.com/facebookexperimental/Recoil/blob/708fcfe93a32beab35b6a9345a05dd35ab736a08/typescript/index.d.ts#L153-L168) `cachePolicy_UNSTABLE.quality = 'value'` option does, as the CodeSandbox demonstrates.

So, I think we have 2 feature requests here:

1. Yours, which I suggest we rename to "Customizable control over cache equality`. This is for the use case where a selector's `get` returns a literal, which Recoil needs help with to tell if it should be considered a new value or not. I suggest we implement this by:
    1. Unhiding `quality`, and if needed...
    2. Allow the value to be a function that receives the old and new value and is expected to return `true` when they should be considered equal.
2. Mine, which I'll create a new issue for to be named "Customizable control subscriptions between values". This is for the use case where a selector's `get` depends on the values of (likely multiple) other selectors or atoms, but doesn't want to subscribe to some of them.

TL;DR for the original / your use case I agree we should go with option 1, and for mine with option 3.

**Comment by RPisRyan** [comment]
I am experimenting with a couple of extensions that might serve as a workaround. Interested to know if there are any obvious problems.
[Demo](https://codesandbox.io/s/recoil-equalatom-and-equalselector-ns1pkv?file=/src/App.tsx)

`
/**
 * Use a writable selector to prevent excess renders.
 * If the setting value is equal to the current value, don't change anything.
 */
export function equalAtom<T>(options: EqualAtomOptions<T>): RecoilState<T> {
  const { key, equals, ...innerOptions } = options;
  const inner = atom({
    key: `${key}_inner`,
    ...innerOptions
  });

  return selector({
    key,
    get: ({ get }) => get(inner),
    set: ({ get, set }, newValue) => {
      const current = get(inner);
      if (!equals(newValue as T, current)) {
        set(inner, newValue);
      }
    }
  });
}
`

`
/**
 * Use a wrapper selector to prevent excess renders.
 * If the latest selection is value-equal to prior ref, return the prior ref.
 */
export function equalSelector<T>(
  options: EqualSelectorOptions<T>
): RecoilValueReadOnly<T> {
  const inner = selector({
    key: `${options.key}_inner`,
    get: options.get
  });

  let prior: T | undefined;

  return selector({
    key: options.key,
    get: ({ get }) => {
      const latest = get(inner);
      if (prior != null && options.equals(latest, prior)) {
        return prior;
      }
      prior = latest;
      return latest as T;
    }
  });
}
`

**Comment by rmill777** [comment]
@FokkeZB For the moment, 	```
cachePolicy_UNSTABLE: {
		equality: 'value'
	}
```is the best solution for me and seems to work how I think selector re-renders should work. Would like to see this feature made stable.

**Comment by drarmstr** [comment]
Note that you can also use a `selectorFamily()` to split up the selector to limit re-evaluations.  Family parameters are serialized for value-equality checks.

**Comment by BenjaBobs** [comment]
That would have a negative effect on the memory footprint though, right?

**Comment by drarmstr** [comment]
It will be cached either way.  You can [adjust the cache policy](https://recoiljs.org/docs/api-reference/core/selector#cache-policy-configuration).  Currently selectors have independent cache policies; it might be interesting to enable selectors to share an LRU cache...

**Comment by FokkeZB** [comment]
@rmill777 @drarmstr be aware that the combination of `cachePolicy_UNSTABLE: { equality: 'value' }` and an async selector dependency that might throw or return a function is at risk of causing https://github.com/facebookexperimental/Recoil/issues/1804

**Comment by drarmstr** [comment]
@FokkeZB Functions can't really be used for value equality because they can contain closure state which isn't easily captured for serialization and comparison.

**Comment by IDisposable** [comment]
> ```ts
> /**
>  * Use a wrapper selector to prevent excess renders.
>  * If the latest selection is value-equal to prior ref, return the prior ref.
>  */
> export function equalSelector<T>(
>   options: EqualSelectorOptions<T>
> ): RecoilValueReadOnly<T> {
>   const inner = selector({
>     key: `${options.key}_inner`,
>     get: options.get
>   });
> 
>   let prior: T | undefined;
> 
>   return selector({
>     key: options.key,
>     get: ({ get }) => {
>       const latest = get(inner);
>       if (prior != null && options.equals(latest, prior)) {
>         return prior;
>       }
>       prior = latest;
>       return latest as T;
>     }
>   });
> }
> ```

Would be awesome to short-circuit when the reference is known to be the same and not bother calling the `options.equals` so that boilerplate doesn't need to be in any of implementations. I'm suggesting:
```ts
const latest = get(inner);
if (prior != null && (lastest !== prior ||  options.equals(latest, prior))) {
    return prior;
}
```

**Comment by IDisposable** [comment]
> ```ts
> ...
>   const priors = new Map<string, T>();
> ... 
>       if (priors.has(priorKey)) {
>         const prior = priors.get(priorKey) as T;
>         if (priors === latest || options.equals(latest, prior)) { // added reference equality check here, it's free
>           return prior;
>         }
>       }
>       priors.set(priorKey, latest);
> ```

As a newbie to this entire library, this may be stupid-obvious, but my long use of OWL/MFC/COM/WinForms/etc. makes me immediately concerned for the bloat in this `Map`. What happens if the keys go stale and how do we manage to purge them?

**Comment by FokkeZB** [comment]
@IDisposable good idea to do a simple equality check first, although that is also what Lodash' `equals` does, which our internally used version of the above defaults to.

To avoid `Map` from becoming too big, you could use a LRU cache: https://www.npmjs.com/package/lru-cache

**Comment by IDisposable** [comment]
Since the quality of the `options.equals` implementation is unknown, and the reference check is trivially cheap, I would strongly suggest that be in the base code.

**Comment by nichita-pasecinic** [comment]
I'd love to see this feature available, thought that's very prioritar for a react state management solution 😄
